uaclient/event_logger.py

```diff
@@ -0,0 +1,152 @@
+"""
+This module is responsible for handling all events
+that must be raised to the user somehow. The main idea
+behind this module is to centralize all events that happens
+during the execution of UA commands and allows us to report
+those events in real time or through a machine-readable format.
+"""
+
+import enum
+import json
+import sys
+from typing import Dict, List, Optional, Set  # noqa: F401
+
+JSON_SCHEMA_VERSION = 0.1
+_event_logger = None
+
+
+def get_event_logger():
+    global _event_logger
+
+    if _event_logger is None:
+        _event_logger = EventLogger()
+
+    return _event_logger
+
+
+@enum.unique
+class EventLoggerMode(enum.Enum):
+    """
+    Defines event logger supported modes.
+    Currently, we only support the cli and machine-readable
+    mode. On cli mode, we will print to stdout/stderr any
+    event that we receive. On machine-readable mode, we will
+    store those events and parse them for an specified
+    format.
+    """
+
+    CLI = object()
+    MACHINE_READABLE = object()
+
+
+class EventLogger:
+    def __init__(self):
+        self._error_events = []  # type: List[Dict[str, Optional[str]]]
+        self._warning_events = []  # type: List[Dict[str, Optional[str]]]
+        self._processed_services = []  # type: List[str]
+        self._failed_services = set()  # type: Set[str]
```

Comment:
Why is this one a `set` while the `processed_services` is a list? I would expect them to be the same

Reply:
Good point